File: kadane subarray sum/kadane.py
from functools import reduce

def max_subarray_sum(arr):
    # Using reduce to process the array iteratively
    res, max_ending = reduce(
        lambda acc, num: (
            max(acc[0], max(acc[1] + num, num)),  # Update global max
            max(acc[1] + num, num)               # Update local max
        ),
        arr[1:],  # Start from second element, as first is pre-initialized
        (arr[0], arr[0])  # Initial values: (global max, local max)
    )
    return res

# Test case
arr = [2, 3, -8, 7, -1, 2, 3]
print("Max Subarray Sum:", max_subarray_sum(arr))

File: kadane subarray sum/kadane.scala
object kadanes extends App{
def maxSubarraySum(arr: Array[Int]): Int = {
  arr.tail.foldLeft((arr(0), arr(0))) { case ((res, maxEnding), num) =>
    val newMaxEnding = Math.max(maxEnding + num, num) // Local max
    val newRes = Math.max(res, newMaxEnding)         // Global max
    (newRes, newMaxEnding)
  }._1
}

// Test case
val arr = Array(2, 3, -8, 7, -1, 2, 3)
println("Max Subarray Sum: " + maxSubarraySum(arr))
}